### ISSUE TITLE
Can the following be achieved by Reloaded.Hooks?

### ISSUE THREAD

**Issue by aybe** [Reloaded-Project/Reloaded.Hooks#17]
Here's some context:

I'd like to see if I can write an OpenGL renderer for [Destruction Derby S3 ViRGE version](https://destructionderby.fandom.com/wiki/Destruction_Derby#S3_Virge_version).

The game is as follows:
- DD.EXE (game executable, calls AYAPI.DLL)
- AYAPI.DLL (their abstraction around S3 API, calls S3DTKW.DLL)
- S3DTKW.DLL (implementation that renders to the [Diamond Stealth 3D 2000](https://en.wikipedia.org/wiki/Diamond_Multimedia#Stealth))

What I'd like to achieve:

Hook onto each exports of both AYAPI.DLL and S3DTKW.DLL so that:
- I can log and understand what's happening between them
- pass on the call to the original DLL to continue execution

In short, some sort of instrumentation.

In [this example](https://github.com/Sewer56/CriFsHook.ReloadedII/blob/master/CriFsHook.ReloadedII/CriHook.cs), I can see how calls are hooked, the thing I haven't figured out is, how is the call forwarded back to the originally hooked method?

Any clues are welcome, thank you!

**Comment by Sewer56** [comment]
Invoke `IHook.OriginalFunction` as per [example in wiki](https://reloaded-project.github.io/Reloaded.Hooks/GettingStarted/#example-file-monitor).

Misc Note: 
Since Graphics APIs are high frequency, consider using [Function Pointers](https://reloaded-project.github.io/Reloaded.Hooks/FunctionPointers/) for your hooks to reduce overhead. This should reduce transition overhead from native to managed to around ~36 x86 instructions (native to managed) transition into your hook. 

Misc Note 2: If absolute performance is a must, the best option is writing a shim in native code for the native graphics API DLLs (like d3d8to9, dxvk do etc,).

**Comment by aybe** [comment]
Thank you for the tips, they'll come in handy at some point!

I have a small problem, I cannot get the Just-In-Time Debugger window to show up.

- the mod has been compiled in Debug configuration
- in the Manage Mods section, the executable is checked for that mod
- in the Configure Mods section, the mod is checked

Things appear to be working in the console:

```
[Reloaded] Create Loader | Time: 114ms
[Reloaded] Loaded: reloaded.sharedlib.hooks in 21ms
[Reloaded] Current Process Location: C:\DERBYS3\DD.EXE
[Reloaded] Loaded: DestructionDerbyS3ViRGE in 106ms
[Reloaded] Loading Mods (Total) | Time: 133ms
[Reloaded] Setting Up Hooks (Async) | Time: 253ms
[Reloaded] Total Loader Initialization Time: 422ms
```

I've reset Visual Studio settings to default, nothing.

Then went to Options, Tools, Debugging, Just-In-Time, found that:.

<img width="526" alt="devenv_z3gRnCRhMS" src="https://user-images.githubusercontent.com/1537591/204781256-95470027-6337-425e-8756-65c29e8aa072.png">

Restarted as administrator so that I can enable it for everything:

<img width="526" alt="devenv_iBooho3hN6" src="https://user-images.githubusercontent.com/1537591/204781133-393f99f3-8ff5-4f9b-a116-9ed1b6042a64.png">

But still nothing, the JIT Debugger Wizard doesn't want to show up.

Any ideas?

Thank you 😀

**Comment by Sewer56** [comment]
Pretty strange and definitively a first I've seen, usually this works out of the box with no configuration required.

[I'd usually refer to this MSDN page for guidance regarding this stuff.](https://learn.microsoft.com/en-us/visualstudio/debugger/debug-using-the-just-in-time-debugger?view=vs-2022)

You can try setting the defaults I suppose: 
in 
- `Computer\HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Windows NT\CurrentVersion\AeDebug`  
and `Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\AeDebug`

Default value of `Debugger` is `"C:\Windows\system32\vsjitdebugger.exe" -p %ld -e %ld -j 0x%p`.
Might work if you manually restore it, not sure as I've never ran into this personally before.

**Comment by aybe** [comment]
I repaired VS, it appeared once for whatever unit tests on some project, I then tried on the project but nothing. 🤬
Been getting all sort of funny stuff lately, e.g. __debugbreak on F12 on a game I'm working on while there's no such code.
I think I'm off for a good old reinstallation, hopefully I'll be able to get started hooking by this weekend.
Anyway, thanks for your help! 😀
